2022/Q2/artifacts/pos-model/PoS-model-cubic.md

```diff
@@ -0,0 +1,493 @@
+func update_total_voting_power(epoch)
+{
+  var total = 0
+  forall (validator in validator_sets[epoch].active U validator_sets[epoch].inactive) do
+    total += validator.voting_power
+  total_voting_power[epoch] = total
+}
+```
+```go
+func update_validator_sets(validator_address, epoch, power_before, power_after)
+{
+  var min_active = first(validator_sets[epoch].active)
+  var max_inactive = last(validator_sets[epoch].inactive)
+  if (power_before >= max_inactive.voting_power) then
+    //if active but it loses power below the max_inactive then move validator to inactive
+    //and promote max_inactive
+    if (power_after < max_inactive.voting_power) then
+      remove(validator_sets[epoch].active, validator_address)
+      add(validator_sets[epoch].active, max_inactive)
+      remove(validator_sets[epoch].inactive, max_inactive.validator)
+      add(validator_sets[epoch].inactive, <validator_address, power_after>)
+    //if active and its change in power does not degrade it to inactive, then update its voting power
+    else
+      remove(validator_sets[epoch].active, validator_address)
+      add(validator_sets[epoch].active, <validator_address, power_after>)
+  else
+    //if inactive and gains power above the min_active, then insert into active and
+    //degrade min_active 
+    if (power_after > min_active.voting_power) then
+      remove(validator_sets[epoch].inactive, validator_address)
+      add(validator_sets[epoch].inactive, min_active)
+      remove(validator_sets[epoch].active, min_active.validator)
+      add(validator_sets[epoch].active, <validator_address, power_after>)
+    //if inactive and its change in power does not promote it to active, then update its voting power
+    else
+      remove(validator_sets[epoch].inactive, validator_address)
+      add(validator_sets[epoch].inactive, <validator_address, power_after>)
+}
+```
+
+```go
+func remove_from_validator_sets(validator_address)
+{
+  var max_inactive = last(validator_sets[epoch].inactive)
+  remove(validator_sets[epoch].active, validator_address)
+  remove(validator_sets[epoch].inactive, max_inactive.validator)
+  add(validator_sets[epoch].active, max_inactive)    
+}
+```
+
+```go
+func is_validator(validator_address, epoch){
+    return read_epoched_field(validators[validator_address].state, epoch, ⊥) == candidate
+}
+```
+
+```go
+end_of_epoch()
+{
+  set_validators = {val | val = slash.validator && slash in enqueued_slashes[cur_epoch]}
+  forall (validator_address in set_validators) do
+    //jail validator (Step 1.2 of cubic slashing). This also unfreeze the validator (Step 2.5 of cubic slashing)
+    validators[validator_address].state[cur_epoch+1] = jailed
+    remove_from_validator_sets(validator_address, cur_epoch+1)
+    //iterate over all slashes for infractions within (-1,+1) epochs range (Step 2.1 of cubic slashing)
+    var set_slashes = {s | s in enqueued_slashes[epoch] && cur_epoch-1 <= epoch <= cur_epoch+1 && s.validator == validator_address}
```

Comment:
As we are iterating over the previous and next epochs, and we will repeat this step for` cur_epoch + 1`, are we not punishing the validator multiple times for the same evidence. So for evidence from cur_epoch, the validator is punished in epochs` cur_epoch -1`, `cur_epoch `and `cur_epoch + 1`. 


Related to this, at the end of` curr_epoch`, we decrease the voting power (and the total voting power). But then at the end of` curr_epoch + 1 `we will do the same thing taking into account an at that point invalid voting power and decreasing it again.

Reply:
> As we are iterating over the previous and next epochs, and we will repeat this step for cur_epoch + 1, are we not punishing the validator multiple times for the same evidence. So for evidence from cur_epoch, the validator is punished in epochs cur_epoch -1, cur_epoch and cur_epoch + 1.

I think I made the same comment during one meeting. This is intended though. The way I currently interpret this is that any evidence may contribute to multiple slashes. So for instance, if there is evidence for a validator at epochs cur_epoch and cur_epoch + 1, the evidence at cur_epoch contributes to the slash rate of the slash computed at cur_epoch + 1 and vice-versa. Once of the goals of the slashing mechanism is to severe punishing to validators that misbehave in consecutive epochs.

Reply:
> Related to this, at the end of` curr_epoch`, we decrease the voting power (and the total voting power). But then at the end of`curr_epoch + 1`we will do the same thing taking into account an at that point invalid voting power and decreasing it again.

I think my previous answer may clarify this. Let me know if it does not.

Reply:
Yes, my understanding of what Heliax want (we discussed it in one of the last meetings) matches Manu's. I believe they are aware of the implications that @jmalicevic points out in her comment above